### PR TITLE
Remove markdown, use markd shard in compiler

### PR DESCRIPTION
**Pull request by RX14**

### CODE REVIEW

**Comment by luislavena** [comment]
@RX14 thank you for submitting this and getting the conversation started.

A few points I noticed:

- Stability: Markd has no public release (yet)
- Performance: own readme indicates is way slower than Crystal built-in version, perhaps worth some peer-review for performance bottlenecks?
- Perhaps this should be listed as development dependencies instead, in case someone decides to use Crystal as a shard and end having to depend on Markd too.

Cheers.

File: src/compiler/crystal/tools/doc/markdown_doc_renderer.cr [hunk 1]
```diff
@@ -38,13 +20,15 @@ class Crystal::Doc::MarkdownDocRenderer < Markdown::HTMLRenderer
...
-        super
+        tag("code")
+        lit(text)
+        tag("/code")
         return
       end
     end
 
     # Check Type#method(...) or Type or #method(...)
+    # https://www.youtube.com/watch?v=SoIfcbsV13w
```
Comment (oprypin) [inline src/compiler/crystal/tools/doc/markdown_doc_renderer.cr]: srsly

Comment (RX14) [reply src/compiler/crystal/tools/doc/markdown_doc_renderer.cr]: Oops, that was a joke and i intended to remove it before I pushed.

But the code below probably does need some refactoring...

**Comment by RX14** [comment]
@luislavena Most of the conversation around whether or not to do this is in #4613, it's be best to keep the discussion there.

It's listed in dependencies because the crystal compiler is actually part of the stdlib and can be required from any crystal code. It's needed at compile time, not just for specs so it's a real dependency.

**Comment by oprypin** [comment]
I don't think "[Improve Markdown implementation](https://github.com/crystal-lang/crystal/issues/4613)" is a suitable title to discuss what's happening here.

Adding a random external dependency to the compiler has so many obvious downsides that I didn't think it would need to be said.

It gives too much control to the owner of that shard. (At the very least move it to crystal-lang org and give the original author non-force push access)

It greatly complicates packaging and installation. Obviously a circular dependency of Crystal -> Markd -> Shards -> Crystal is very bad (actually I don't quite see how to even resolve it, since Shards is not even a dependency of Crystal). But this also makes Crystal extremely inaccessible for distribution in many forms, I'm mainly thinking of Linux distributions right now, for example, they would have to make an unprecedented move of adding a Crystal library to the distro, AND introduce a circular dependency of Crystal <-> Markd. Enterprise users would not be happy about such a dependency either. And the mixed licenses wouldn't help.

**Comment by oprypin** [comment]
...
Bootstrapping becomes harder for no reason.

If you're gonna be vendoring the dependency, you must do it properly, e.g. by copying the source code or directly downloading it. `shards` doesn't have to be a part of this.

**Comment by asterite** [comment]
Another idea would be to move the documentation generator to a completely separate project (and binary). I was wondering how Elixir did it for docs, because they also use markdown, and it turns out it's a separate project indeed: https://github.com/elixir-lang/ex_doc (last commit is from José Valim, so it's also a core project). The doc generate needs the compiler as a dependency, but that's fine. And it's a less critical tool than the compiler itself.

This also means the compiler doesn't need to know about docs, nor use ECR. We could probably keep extracting these tools to different binaries. The only downside is that every binary will be kind of big, because the compiler's code is big. Maybe 20MB each executable. Maybe not a big problem.

**Comment by asterite** [comment]
In fact it seems `ex_doc` can depend on different markdown implementations, which is also great because no more "But I want to use this markdown implementation because it has feature X".

**Comment by akzhan** [comment]
I approve  the idea but can't approve embedding of shard that owned not by crystal-lang community.

**Comment by icyleaf** [comment]
Sorry reply late, i'm glad `markd` may be or will be a part of crystal in anyways.
It is welcome to  merged into crystal or move to crystal-lang group.

**Comment by asterite** [comment]
@icyleaf I would really like to include your code in the standard library, but https://github.com/icyleaf/markd/pull/7 made the performance terrible so now that's not possible. We should fix the performance problems and then include it in the std.

**Comment by asterite** [comment]
Actually, it's https://github.com/icyleaf/markd/pull/5 that made everything slower

**Comment by akzhan** [comment]
@asterite May you move `markd` to crystal-lang organization? All things be simpler with it.

**Comment by asterite** [comment]
@akzhan No, because to compile crystal we will then needs shards, which was said to be very difficult.

**Comment by straight-shoota** [comment]
The immediate performance hit was easily resolved (icyleaf/markd#8). And there are probably plenty of opportunities to improve performance significantly. So far, I have only worked on some of the most outstanding issues with code quality. There are still lot's of Javascriptisms in the code - it needs to be *crystalized* ;)

Currently, performance of markd is about 9 times slower than stdlib implementation (based on markd's `benchmar/source.md`) though that comparison is not fair because `Markdown` simply neglects some markdown features.

**Comment by RX14** [comment]
This is a stale PR at best

**Pull request closed without merging**